docs/Extensibility.md

```diff
@@ -0,0 +1,104 @@
+# Extensibility
+
+> Expect this pattern to evolve (as it has already done) or even disappear if [proper extensibility features](https://community.starknet.io/t/contract-extensibility-pattern/210/11?u=martriay) are implemented into Cairo.
+
+* [The extensibility problem](#the-extensibility-problem)
+* [The pattern ™️](#the-pattern)
+  * [Libraries](#libraries)
+  * [Contracts](#contracts)
+* [Presets](#presets)
+* [Emulating hooks](#emulating-hooks)
+
+## The extensibility problem
+
+Smart contract development is a critical task. As with all software development, it is error prone; but unlike most scenarios, a bug can result in major losses for organizations as well as individuals. Therefore writing complex smart contracts is a delicate task.
+
+One of the best approaches to minimize introducing bugs is to reuse existing, battle-tested code, a.k.a. using libraries. But code reutilization in StarkNet’s smart contracts is not easy:
+
+- Cairo has no explicit smart contract extension mechanisms such as inheritance or composability
+- There’s no function overloading nor aliasing which makes function selector collisions very likely – more so considering selectors do not take function arguments into account
+- Any `@external` function defined in an imported module will be automatically re-exposed by the importer (i.e. the smart contract)
+
+To overcome these problems, this project builds on the following guidelines™.
+
+## The pattern
+
+The idea is to have two types of Cairo modules: libraries and contracts. The flow then is for libraries to define reusable logic and storage variables which can then be extended and exposed by contracts. Contracts can be deployed, libraries cannot.
+
+To minimize risk, boilerplate, and avoid semantic clashes, we follow these rules:
+
+### Libraries
+
+- All function and storage variable names must be prefixed with the file name to prevent clashing with other libraries (e.g. `ERC20_approve` in the `ERC20` library)
+- Must not implement any `@external` or `@view` functions
+- Must not implement constructors
+- Must not call initializers on any function
+- Should implement initializers if needed (as any other library function, never as `@external`)
+
+### Contracts
+
+- Can import from libraries
+- Should implement `@external` functions if needed
+- Should implement a constructor that calls initializers
+- Must not call initializers on any function aside the constructor
+
+Note that since initializers will never be marked as `@external` and they won’t be called from anywhere but the contract constructor, there’s no risk of double-initialization. It’s up to the library developers not to make initializers inter-dependent to avoid weird dependency paths.
```

Comment:
This is contradictory. There is risk of double initialization.

```suggestion
Note that since initializers will never be marked as `@external` and they won’t be called from anywhere but the contract constructor, there’s no risk of re-initialization after deployment. It’s up to the library developers not to make initializers interdependent to avoid weird dependency paths that may lead to double initialization of libraries.
```